webhook/credentials.go

```diff
@@ -0,0 +1,205 @@
+	serviceAccount := &corev1.ServiceAccount{}
+	namespacedName := types.NamespacedName{
+		Name:      pod.Spec.ServiceAccountName,
+		Namespace: pod.Namespace,
+	}
+
+	err := w.client.Get(ctx, namespacedName, serviceAccount)
+	if err != nil {
+		return nil, err
+	}
+
+	return serviceAccount, nil
+}
+
+func (w *CredentialsInjector) getLogger(ctx context.Context) logr.Logger {
+	logger := log.FromContext(ctx)
+	return logger.WithName("credentials-injector-webhook")
+}
+
+func getPatchedResponse(req admission.Request, mutatedPod *corev1.Pod) admission.Response {
+	marshaledPod, err := json.Marshal(mutatedPod)
+	if err != nil {
+		return admission.Errored(http.StatusInternalServerError, err)
+	}
+
+	return admission.PatchResponseFromRaw(req.Object.Raw, marshaledPod)
+}
+
+func injectEnvVar(container *corev1.Container) {
+	credentialsPath := fmt.Sprintf("%s/%s", VolumeMountWorkloadIdentityPath, GoogleApplicationCredentialsJSONPath)
+
+	credentialsEnvVar := corev1.EnvVar{
+		Name:  EnvKeyGoogleApplicationCredentials,
+		Value: credentialsPath,
+	}
+	container.Env = append(container.Env, credentialsEnvVar)
+}
+
+func injectVolumeMount(container *corev1.Container) {
+	credentialsMount := corev1.VolumeMount{
+		Name:      VolumeWorkloadIdentityName,
+		MountPath: VolumeMountWorkloadIdentityPath,
+		ReadOnly:  true,
+	}
+	container.VolumeMounts = append(container.VolumeMounts, credentialsMount)
+}
+
+func injectVolume(pod *corev1.Pod, workloadIdentityPool, configMapName string) {
+	pod.Spec.Volumes = append(pod.Spec.Volumes, corev1.Volume{
+		Name: VolumeWorkloadIdentityName,
+		VolumeSource: corev1.VolumeSource{
+			Projected: &corev1.ProjectedVolumeSource{
+				DefaultMode: int32ptr(VolumeWorkloadIdentityDefaultMode),
+				Sources: []corev1.VolumeProjection{
+					{
+						ServiceAccountToken: &corev1.ServiceAccountTokenProjection{
+							Path:     ServiceAccountTokenPath,
+							Audience: workloadIdentityPool,
+
+							// According to documentation, the service account token will be
+							// rotated automatically by the kubelet when it's close to
+							// expiring.
+							// See https://kubernetes.io/docs/tasks/configure-pod-container/configure-service-account/#service-account-token-volume-projection
+							ExpirationSeconds: int64ptr(TokenExpirationSeconds),
+						},
+					},
+					{
+						ConfigMap: &corev1.ConfigMapProjection{
+							LocalObjectReference: corev1.LocalObjectReference{
+								Name: configMapName,
+							},
+							Optional: boolptr(false),
+							Items: []corev1.KeyToPath{
+								{
+									Key:  ConfigMapKeyGoogleApplicationCredentials,
+									Path: GoogleApplicationCredentialsJSONPath,
+								},
+							},
+						},
+					},
+				},
+			},
+		},
+	})
+}
+
+func int32ptr(i int) *int32 {
+	u := int32(i)
+	return &u
+}
+
+func int64ptr(i int) *int64 {
+	u := int64(i)
+	return &u
+}
+
+func boolptr(b bool) *bool {
+	return &b
+}
```

Comment:
Not essential but we have a module to handle this: https://github.com/giantswarm/to

Reply:
Awesome! That's pretty useful. I've switched it to use that instead, thanks